[WIP] Progressive call invocations (#453)
wamp-proto/wamp-proto: rfc/text/advanced/ap_10_rpc_progressive_call_invocations.md

## Progressive Call Invocations {#rpc-progressive-call-invocations}

A *Caller* may issue a progressive call with progressive call invocations. This can be useful in a few cases:

* Payload is too big to send it whole in one request, e.g., uploading a file.
* Long-term data transfer that needs to be consumed early, such as a media stream.
* RPC is called too often and overall processing can be optimized: avoid the need to generate a new id for requests,
  initiate data structures for a new call, etc.

In such cases, a procedure implemented by a *Callee* and registered at a *Dealer* may be made to receive progressive call invocations,

> where the *Callee* waits after receiving the entire set of payload chunks before sending the result.

Reviewer: Looking at this sentence, I start thinking that we should rewrite it. As `waits after receiving the entire set of payload` is not the only possible way. And *Callee* may start processing data ASAP. And even return chunks of results (with the help of progressive call results)

Reviewer 2: We are describing the Progressive Call Invocations feature independently of Progressive Call Results at this point. The new *Progressive Calls with Progressive Call Results* subsection I consolidated discussed the mixing of both.

Author: what is a "payload chunk"? this term isn't defined or required

> where the *Callee* waits after receiving the entire set of payload chunks before sending the result.

Reviewer: ```suggestion
where the *Callee* may start processing the incoming data without awaiting the entire set of payload chunks.
```

Reviewer: @oberstet @ecorm What do you think of this change?

Author: it seems the above 2 paragraphs are to describe the 2 options a callee that receives progressive call invocations (not "payload chunks") does have. start processing _and_ start producing (progressive) results immediately OR start receving progressive invocations and processing (or not) immediately, but continue to do so before finally producing 1 non-progressive results.

if so, I agree: it makes sense to explicitly describe those differing behaviors possible for a callee

Reviewer: agree! 👍🏻

Reviewer 2: > What do you think of this change?

LGTM

## Progressive Call Invocations {#rpc-progressive-call-invocations}

> A *Caller* may issue a progressive call having progressive invocations. This can be useful in a few cases:

Reviewer 2: ```suggestion
A *Caller* may issue a call having progressive invocations. This can be useful in a few cases:
```

> The following cases, where a *Caller* sends a `CALL` message with `progress := true`, MUST be treated as *protocol errors*

Reviewer 2: ```suggestion

As a consequence, *Dealers* MUST also announce support of the _Call Cancelling_ feature via

{align="left"}
        WELCOME.Details.roles.dealer.features.call_cancelling|bool := true

The following cases, where a *Caller* sends a `CALL` message with `progress := true`, MUST be treated as *protocol errors*
```

Author: -1

    ]
```

A call invocation MUST *always* end in a *normal* `CALL` without the `"progress": true` option, or explicitly
set `"progress": false` which is the default.

**Progressive Call Invocations and Shared Registration**

RPCs can have a multiple registrations (see `Shared Registration` feature) with different `<invocation_policies>`.
However, allowing progressive `CALL` messages to be routed to different *Callees* would lead to unexpected results.

> To prevent this the *Dealer* must make a first `INVOCATION` based on `<invocation_policy>` and then route all

Reviewer 2: ```suggestion
To prevent this the *Dealer* MUST make a first `INVOCATION` based on `<invocation_policy>` and then route all
```

        CALL.Options.sticky|bool := true

In this case, the *Dealer* must make a first `INVOCATION` based on `<invocation_policy>` and then route all
subsequent `progressive` calls to the same *Callee*.

> If binding all ongoing `progressive` calls to the same *Callee* is not required, the *Caller* can set the `sticky` option to `false`.

Reviewer 3: Hmmm this part was removed right? Or is it because this change is based on an older version of the docs?

Reviewer: kinda Yes and No :)
The original PR with removal was merged but then due to conflicts was reverted. So we have an opened #460 once again. But it needs to be merged after this one. And then — everything should be all good %)

Reviewer: FYI @Jopie64 Merged PR with sticky :)

                         `------'                 `------'

Note: Any `ERROR` returned by the *Callee*, in response to an `INTERRUPT`, is ignored (same as in regular call
canceling when mode="killnowait"). It is therefore not necessary for the *Callee* to send an `ERROR` message.

**Ignoring Progressive Call Invocations**

Unlike some other advanced features, a *Callee* cannot be unaware of progressive call invocations.
Therefore, if a *Callee* doesn't support this feature, the *Dealer* MUST respond to the *Caller* with an
`wamp.error.feature_not_supported` error message.

A *Callee* that supports progressive call invocations, but does not support call canceling, shall be considered by the *Dealer*
as not supporting progressive call invocations.

**Progressive Call Invocations with Progressive Call Results**

*Progressive Call Invocations* may be used in conjunction with *Progressive Call Results* if the *Caller*, *Dealer*, and *Callee* all support both features. This allows the *Callee* to start sending partial results back to the *Caller* after receiving one or more initial payload chunks. Efficient two-way streams between a *Caller* and *Callee* can be implemented this way.

> The following message flow illustrates a progressive call using progressive call invocations when a *Callee* starts sending progressive call

Reviewer 2: ```suggestion
The following message flow illustrates a call using progressive call invocations when a *Callee* starts sending progressive call
```
